python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -750,15 +773,37 @@ class PowerSupplyOrFan(Shadow):
     __shadowclass__ = manage.PowerSupplyOrFan
     __lookups__ = [('netbox', 'name')]
 
+    def __init__(self, *args, **kwargs):
+        super(PowerSupplyOrFan, self).__init__(*args, **kwargs)
+        self.is_new = None
+
     def prepare(self, containers):
-        existing = self.get_existing_model(containers)
+        self.is_new = not self.get_existing_model()
         # Set a default value of UNKNOWN if this is a new object
-        if not existing and self.up is None:
+        if self.is_new and self.up is None:
             self.up = manage.PowerSupplyOrFan.STATE_UNKNOWN
 
+    def cleanup(self, containers):
+        self._handle_new_psu_or_fan()
+
+    def _handle_new_psu_or_fan(self):
+        if self.is_new:
+            psufan = self.get_existing_model()
+            try:
+                device_event.notify(
+                    device=psufan.device,
+                    netbox=psufan.netbox,
+                    alert_type="deviceNewPsu" if psufan.is_psu() else "deviceNewFan",
+                ).save()
+            except manage.Device.DoesNotExist:
+                self._logger.debug(
+                    f"New PowerSupplyOrFan does not have a Device defined."
+                )
```

Comment:
Is it not possible to first check if the power supply or fan has a device defined before trying to create an event?

Reply:
Django does indeed raise `DoesNotExist` if attempting to access or lookup some database object that doesn't exist.

A simpler method might be to change line 790 to something like `if self.is_new and self.device` and skip the exception handler - because, yes, a PSU or Fan doesn't necessarily correspond to a device. A Device entry isn't very interesting unless we have a serial number for something, and it's not given that any MIB that provides information about fans and power supplies necessarily provides us with serial numbers for them (sometimes the MIB sucks, sometimes these components just aren't field replaceable but are still reported for status).